File: internal/eligibility/eligibility.go
package eligibility

import (
	"driver-order-allocation-simulator/internal/model"
)

func IsEligible(driver model.Driver, order model.Order, mode string) bool {
	if !driver.Online {
		return false
	}
	if driver.AccountStatus != "active" {
		return false
	}

	supportsService := false
	for _, svc := range driver.ServiceTypes {
		if svc == order.ServiceType {
			supportsService = true
			break
		}
	}
	if !supportsService {
		return false
	}

	if mode == "hard" && driver.DeviceStatus != "healthy" {
		return false
	}

	return true
}

func FilterEligible(drivers []model.Driver, order model.Order, mode string) []model.Driver {
	var result []model.Driver
	for _, d := range drivers {
		if IsEligible(d, order, mode) {
			result = append(result, d)
		}
	}
	return result
}

File: src/eligibility.py
from typing import List
from .models import Driver, Order

def is_eligible(driver: Driver, order: Order, mode: str = "hard") -> bool:
    """Check if driver is eligible for the order.
    
    Hard constraints:
    - Must be online
    - Account must be active
    - Must support the service type
    
    In 'hard' mode, device_status must be healthy.
    In 'soft' mode, device_status is not a filter (handled in scoring).
    """
    if not driver.online:
        return False
    if driver.account_status != "active":
        return False
    if order.service_type not in driver.service_types:
        return False
    if mode == "hard" and driver.device_status != "healthy":
        return False
    return True

def filter_eligible(drivers: List[Driver], order: Order, mode: str = "hard") -> List[Driver]:
    return [d for d in drivers if is_eligible(d, order, mode)]
